### ISSUE TITLE
Since rails 7.1, executing with a disconnected Mysql2Adapter no longer raises ConnectionNotEstablished.

### ISSUE THREAD

**Issue by fumihumi** [rails/rails#49733]
Since rails 7.1, executing with a disconnected Mysql2Adapter no longer raises ConnectionNotEstablished.

### Steps to reproduce
in my project, conn.disconnect! is used to disconnect the connection after detecting a temporary database connection failure.
 (e.g. failover / connection failure due to restart...etc)

Until Rails 7.0.8, when using a disconnected connection, ActiveRecord::ConnectionNotEstablished would be raised and conn.execute would fail.
(I have a mechanism to try to reconnect using this ConnectionNotEstablished Error)

Since Rails 7.1.0, conn.execute does not raise an error despite being disconnected.

Perhaps this change was made by this PR ( https://github.com/rails/rails/pull/44591).

Are there any plans to modify it to raise the error?

```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"
  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", '7.1.1' # failed
  # gem "rails", '7.1.0' # failed
  # gem "rails", '7.0.8' # passed

  gem "mysql2"
end

require "active_record"
require "minitest/autorun"
require "logger"

db_name = "test_connection"
options = { host: "127.0.0.1", port: 3306 }

adapter = 'mysql2'

ActiveRecord::Base.establish_connection(adapter: adapter, username: "root", **options)
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Base.connection.drop_database db_name
ActiveRecord::Base.connection.create_database db_name
ActiveRecord::Base.establish_connection(adapter: adapter, username: "root", database: db_name, **options)
ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
    t.string :name
  end
end

class User < ActiveRecord::Base; end

class BugTest < Minitest::Test
  def test_disconnected_conn_execute
    conn = User.connection
    conn.disconnect!
     # v7.0.8: pass, v7.1.0 and v7.1.1: failed
    assert_raises(ActiveRecord::ConnectionNotEstablished) { conn.execute("select 1;") }
  end
end
```

### Expected behavior

ConnectionNotEstablished is raised in case of execute using a disconnected connection.

### Actual behavior

The execute method does not raise an error.
execute completes successfully and returns Mysql::Result.

### System configuration
**Rails version**: 7.1.1
**Ruby version**: 3.2.2

**Comment by matthewd** [comment]
Explicit `disconnect!` is an edge case, but generally I think this is expected behaviour -- if we aren't connected, we're trying to do something that requires us to be connected, and we know how to connect... it seems reasonable that we'd do the thing that should work.

We _could_ specifically track the fact that we've manually disconnected and therefore should not auto-reconnect, but it mostly just seems like avoidable complication.

I'm open to being convinced, though.. is there a compelling use for being able to leave the connection broken? (You describe error recovery & reconnection code in your app -- does Active Record's new automatic reconnection behaviour not work for you?)

**Comment by byroot** [comment]
I agree with @matthewd  that this sounds more like a "bug fix / improvements" than a regression.

So I'll close but we can re-open if you still consider it a bug and make a case for it.

**Comment by fmichaut-diff** [comment]
I'd like to make a case for it : 
Let's say I need to fetch a bunch of records from the DB. I would most likely use `includes` and other preload methods to ensure it is fetched efficiently. This ensure we fetch everything first, and we shouldn't need to query anything else after that (in theory only tho).

The thing is, when the application grows the code that is using the preloaded records will change, and you might forget to update the `includes`. While that is not always a big deal it can be. 
Our use-case is a huge query that needs to be preloaded before being processed by a bunch of parallel threads. If we forget to preload an association, we might not notice it in Dev (gems like bullet won't help cause its not in Controller/Views) but on prod we will reach `TooManyOpenConnections` because of the huge volume of records/processes we are dealing with.

So I would like to protect the code against ever performing queries in the `Thread` blocks. I wanted to use `disconnect!` before and re-enable after `join`-ing on all the threads in an ensure block - but this doesn't work because of current Rails behaviour.

Basically I want a way to tell ActiveRecord "if anything attempts to run a query in this block -> raise an error".
A method like `ActiveRecord::Base.disable_queries do ... end` would also do just fine. 

TL;DR : I just want a safety net to ensure a block of code cannot reach the DB. There might be a couple of reason to want to do so, hence I think it should be implemented

**Comment by matthewd** [comment]
I can understand that desire, but I don't think I'd want to encourage using true DB disconnection to achieve it -- that might work in a one-shot script, but it'd only be a matter of time before someone copied the pattern into a controller or job context, where you Really Really Should Not hard-disconnect database connections.

(In fact, I wouldn't've expected the previous behaviour to work in this situation either: the original thread's connection would be disconnected, but the pool would still happily allocate a new connection to each requesting thread. You'd have to break the entire pool... which should still "work" today.)

It's not a documented / officially supported behaviour (so no promises it won't break in future), but _within a thread_, I think you might find `ActiveRecord::Base.connected_to(role: :some_role_that_does_not_exist) do .. end` might be a workable spelling of `disable_queries`.

**Comment by byroot** [comment]
Or just use an ActiveSupport::Notifications subscriber on `sql.active_record` and raise if a query is performed and some thread variable is set.
